### PR TITLE
Add 0xApi & Errors

### PR DESCRIPTION
**Pull request by W3stside**
Adds a small 0xApi to fetch quotes

- [x] Price fetching tests
- [x] Price error tests - inside #48

### CODE REVIEW

**Comment by matextrem** [comment]
We need to check coveralls for this PR since it was decreased around a 7%

File: src/0xSdk.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  constructor(chainId: T, cowContext: CowContext = {}, options: Options = {}) {
+    this.context = new Context(chainId, { ...cowContext })
+    this.api = new ZeroXApi(this.context)
+
+    log.setLevel(options.loglevel || 'error')
+  }
+
+  updateChainId = (chainId: ChainId) => {
+    this.context.updateChainId(chainId)
+  }
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/0xSdk.ts:23]: I should actually remove this, it's not used in the 0xAPI - this is some legacy code, good eye

File: src/api/0x/0x.spec.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
+import fetchMock, { enableFetchMocks } from 'jest-fetch-mock'
+import { OrderKind } from '@cowprotocol/contracts'
+import { PriceQuoteParams } from '../cow/types'
+import { SupportedChainId } from '../../constants/chains'
+import { ZeroXSdk } from '../../0xSdk'
+import ZeroXError from './error'
+
+enableFetchMocks()
+
+const chainId = 1 as SupportedChainId //Rinkeby
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/api/0x/0x.spec.ts:10]: :+1

File: src/api/0x/0x.spec.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      amount: '0',
+      kind: OrderKind.BUY,
+    } as PriceQuoteParams)
+  } catch (e) {
+    const error = e as ZeroXError
+
+    expect(error.message).toEqual('Validation Failed')
+    expect(error.description).toEqual('INSUFFICIENT_ASSET_LIQUIDITY')
+    expect(fetchMock).toHaveBeenCalledTimes(1)
+  }
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/api/0x/0x.spec.ts:101]: will do

File: src/0xSdk.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (W3stside) [reply src/0xSdk.ts:23]: i've added 0xApi to the CowSdk directly

File: src/api/0x/error/index.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        return errorMessage
+      } else if (errorResponse?.reason) {
+        log.error(logPrefix, 'Unknown validation reason for bad price request', errorResponse)
+        return errorResponse.reason
+      } else {
+        throw 'Error response body properties "reason" and "validationErrors" missing.'
+      }
+    } catch (error) {
+      log.error(logPrefix, 'Error handling a 4xx error. Likely a problem deserialising the JSON response')
+      return 'Price fetch failed. This may be due to a server or network connectivity issue. Please try again later.'
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/api/0x/error/index.ts:39]: yeah good catch

File: src/api/0x/error/index.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        return 'Internal server error'
+      case 501:
+        return 'Not Implemented'
+      case 503:
+        return 'Server Error - Too many open connections'
+
+      default:
+        log.error(
+          logPrefix,
+          '[ZeroXError::getErrorFromStatusCode] Error fetching quote, status code:',
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/api/0x/error/index.ts:62]: yep good catch pt. 5000

**Review by alfetopito — APPROVED** [review]

File: src/api/0x/index.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+import BaseApi from '../base'
+
+import { Options, PriceQuoteParams } from '../cow/types'
+import ZeroXError, { logPrefix, ZeroXErrorResponse } from './error'
+import { ERC20BridgeSource, MatchaOptions, MatchaPriceQuote } from './types'
+
+// GPV2Settlement
+// https://etherscan.io/address/0x9008d19f58aabd9ed0d60971565aa8510560ab41
+const AFFILIATE_ADDRESS = '0x9008D19f58AAbD9eD0D60971565AA8510560ab41'
```
Comment (ramirotw) [inline src/api/0x/index.ts:16]: why not referencing https://github.com/cowprotocol/cow-sdk/blob/base-api/src/constants/index.ts#L7 ?

Comment (W3stside) [reply src/api/0x/index.ts:16]: cool, didn't know it existed 👍

File: src/api/0x/index.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    case SupportedChainId.MAINNET:
+      return chainId
+
+    default:
+      // Unsupported network
+      return null
+  }
+}
+
+function _optionsToSearchUrl(options: MatchaOptions) {
```
Comment (ramirotw) [inline src/api/0x/index.ts:42]: explicit return type is missing

File: src/api/0x/index.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+  const { excludedSources, affiliateAddress } = options
+
+  const excludedSourceString = excludedSources.length > 0 ? excludedSources.join(',') : ''
+  const optionsMap = Object.entries({ excludedSources: excludedSourceString, affiliateAddress }).map(
+    ([option, value]) => `${option}=${value}`
+  )
+  return optionsMap.join('&')
```
Comment (ramirotw) [inline src/api/0x/index.ts:49]: could this utility function work here? https://github.com/cowprotocol/cow-sdk/blob/44e38ba32db433778836b2d47a55bca61411bfbf/src/utils/common.ts#L10

File: src/api/0x/index.ts [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+async function _handleQuoteResponse<T = unknown, P extends PriceQuoteParams = PriceQuoteParams>(
+  response: Response,
+  params?: P
+): Promise<T> {
+  if (!response.ok) {
+    const errorObj: ZeroXErrorResponse = await response.clone().json()
```
Comment (ramirotw) [inline src/api/0x/index.ts:57]: why is the clone needed here?

File: src/api/0x/index.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+  async getQuote(params: PriceQuoteParams, options: Options = {}): Promise<MatchaPriceQuote | null> {
+    const { amount, baseToken, quoteToken, kind } = params
+    const { chainId: customChainId, isDevEnvironment = this.context.isDevEnvironment } = options
+    const chainId = customChainId || (await this.context.chainId)
+    // this is handled via an error on L51 via the fetch call
+    // but we can handle it here to control the flow better
+    const networkId = _getMatchaChainId(chainId)
+    if (networkId == null) {
+      log.debug('[0x API] - Network not supported')
```
Comment (ramirotw) [inline src/api/0x/index.ts:99]: shouldn't we reuse the logprefix in every log statement?

**Review by ramirotw — APPROVED** [review]

File: src/api/0x/index.ts (hunk 9 above)
Comment (W3stside) [reply src/api/0x/index.ts:57]: because the body is used twice (here and L62)

File: src/api/0x/index.ts (hunk 10 above)
Comment (W3stside) [reply src/api/0x/index.ts:99]: good catch

File: src/api/0x/index.ts (hunk 8 above)
Comment (ramirotw) [reply src/api/0x/index.ts:49]: I see this is concatenated later in:

```js
 `/price?sellToken=${sellToken}&buyToken=${buyToken}&${swapSide}=${amount}&${this.MATCHA_OPTIONS_URL}`
```

and this function should we called just once with every qs parameter. Maybe handling all the parameters as an object?

Comment (W3stside) [reply src/api/0x/index.ts:49]: think so, will adapt my code

**Pull request merged**